src/mlpack/core/optimizers/parallel_sgd/sgdp_impl.hpp

```diff
@@ -0,0 +1,153 @@
+ */
+#ifndef __MLPACK_CORE_OPTIMIZERS_PARALLELSGD_SGDP_IMPL_HPP
+#define __MLPACK_CORE_OPTIMIZERS_PARALLELSGD_SGDP_IMPL_HPP
+
+#include <mlpack/methods/regularized_svd/regularized_svd_function.hpp>
+// In case it hasn't been included yet.
+#include "sgdp.hpp"
+namespace mlpack {
+namespace optimization {
+
+template<typename DecomposableFunctionType>
+ParallelSGD<DecomposableFunctionType>::ParallelSGD(DecomposableFunctionType& function,
+                                   const double stepSize,
+                                   const size_t maxIterations,
+                                   const double tolerance) :   
+    function(function),
+    stepSize(stepSize),
+    maxIterations(maxIterations),
+    tolerance(tolerance)
+{ /* Nothing to do. */ }
+
+
+
+//! Optimize the function (minimize).
+template<typename DecomposableFunctionType>
+double ParallelSGD<DecomposableFunctionType>::Optimize(arma::mat& iterate)
+{
+
+  // Find the number of functions to use.
+  const size_t numFunctions = function.NumFunctions();
+
+  // To keep track of where we are and how things are going.
+  double overallObjective = 0;
+  double lastObjective = DBL_MAX;
+
+  //get maximum number of threads that will be running. with is defined by OMP_NUM_THREADS
+  size_t num_thread=omp_get_max_threads();
+
+  //T: after T iteration it will go for tollerence checking;  we can define T in many way
+  //size_t T=maxIterations/num_thread;
+  //Here I have thaken T=contant
+  size_t T=1000;
+
+
+  //vector of iterate. length of tIterate is same as number of threads available.
+  std::vector<arma::mat> tIterate; 
+
+  //initializing each element of tIterate with initial iterate value 
+  for(size_t i=0;i<num_thread;i++)
+  {
+    tIterate.push_back(iterate);
+  }
+
+  //sumIterate is taken track the  sum  all other computed iterate value from each thread. 
+  arma::mat sumIterate(iterate.n_rows,iterate.n_cols);
+  arma::mat gradient(iterate.n_rows, iterate.n_cols);  //gradient for each thread.
+  size_t it;   
+  bool halt=false;
+  sumIterate.zeros();
+  math::RandomSeed(std::time(NULL));
+
+  
+  #pragma omp parallel  shared(sumIterate,halt) private(it,gradient) 
+  {
+    it=1; 
+    int selectedFunction;
+    int th_num;
+    while(it!=maxIterations && halt != true)
+    {
+      it++;
+
+      th_num=omp_get_thread_num(); //thread number is stored in which the thread is running. 
+      selectedFunction=(int)numFunctions*math::Random();
+      function.Gradient(tIterate[th_num],selectedFunction, gradient);
+      tIterate[th_num] -= stepSize * gradient;
+      
+      //checking whether or not it will go for tollerence checking 
+      if(it%T!=0)
+      {
+        continue;
+      }
+
+
+      #pragma omp critical 
+      {
+        sumIterate += tIterate[th_num];
+      }
+    
+      //wait untill all thread update sumIterate
+      #pragma omp barrier   
+
+
+      //runing  a single thread for tollerence checking
+      #pragma omp master
```

Comment:
I thought that the tolerance checking was done for each algorithm individually.  As in, all you would need to do for this algorithm would be something like:

```
#pragma omp parallel
{
  SGD sgd(...);
  sgd.Optimize(...);
}

// Then combine results.
```

I think that this could be a lot simpler still.  Since we already have the `SGD` class, why not use it? :)  Let me know if I've misunderstood something.